Support includePath as an option.

fuzzybinary opened the pull request:
This allows for use of things like the material_components_web node_module which assumes that the root @material directory is searchable on an include path.

Note that includePaths still need to be readable by the build

fuzzybinary committed:
Support includePath as an option.

This allows for use of things like the material_components_web node_module which assumes that the root @material directory is searchable on an include path.

Note that includePaths still need to be readable by the build

fuzzybinary commented:
It does have to be in either web or lib (I installed in web so I could also access the javascript in the node module through a binding project)

My current setup is to npm install material-components-web in the web directory and add "web/node_modules" to the includePaths.  I can then include / use any of the material components / mixins in my sass module.  If you were using a sass only module, I believe installing to lib would also work.

Let me know if that makes sense.

I do plan on fixing the dart_format issues - I just haven't had time to get back to it yet.

nshahan commented:
In general the idea of being able to map a directory to be included in the search seems like a good idea so that you can get access to sass files that are shared in some way outside of the dart package ecosystem. 

I am a little afraid of the complications this can add for the build. Placing the node_modules directory under web/ in your package will include files in the build that you probably don't want. This could inflate the size of your build graph, invalidate builds unnecessarily, copy files you don't need to the output, etc.

If we add this, I think the best practice here would be to include/exclude files from the build with a build.yaml file with some guidance about how to use this feature in the README.

fuzzybinary committed:
Fix dartfmt issues

fuzzybinary commented:
Inflated build graphs ended up not being an issue in reality, but to be honest I'm not exactly sure why not (this is my ignorance of build / build_runner comming through)

In reality I would love to be able to have sass_builder only add included files to the build graph, and have them hosted outside web/lib, but any location outside there fails the BuildStep::canRead check. Is there a way to bypass that, or should files processed via "includePath" (or an include / exclude pattern matcher) use a different method or resolving file?

luisvt commented:
@fuzzybinary is not possible just to create a pub library that contains the material components?

I mean, you can separate the `material-components` sass into a separate library and maybe try to keep it updated, something like I do with `bootstrap_sass`. I know it's going to be more diffiecult to mantain 2 libraries.

jakemac53 commented:
There are some real concerns regarding the `node_modules` directory and the build system due to its huge potential size.

One way to mitigate this is to create two targets, excluding `node_modules` from the main target and only including the exact files you want in another target.

```yaml
targets:
  $default:
    sources:
      include:
        - lib/**
        - web/**
      exclude:
        - web/node_modules/**
  node_modules:
    sources:
      include:
        - web/node_modules/**.sass
```

However due to how file watching works you might still run into issues with to many file handles being open, and there is no known resolution to that :/

nshahan commented:
After some discussion and time away from this issue, this does seem reasonable even if there are risks with including directories that may bloat the build graph.

@fuzzybinary Thanks for your contribution. Let's add this feature. Could you add a simple description of the new option to the Builder Options section of the README?
https://github.com/dart-league/sass_builder/blob/master/README.md#builder-options

fuzzybinary commented:
@nshahan Yeah, I'll take care of it as soon as I get some time, should be sometime in the next week or so.

fuzzybinary committed:
Add basic documentation for includePaths.

nshahan commented:
@fuzzybinary Thanks! I just merged another change so I'll fix up the conflicts.

nshahan committed:
Merge branch 'master' into feature/includePath

nshahan commented on README.md:
```diff
@@ -98,6 +98,9 @@ To configure options for the builder see the `build_config`
 
 * `outputStyle`: Supports `expanded` or `compressed`. Defaults to `expanded` in
   dev mode, and `compressed` in release mode.
+* `includePaths`: Supports adding a list of  directories that will be searched 
```
nit: There is an extra space here.

nshahan commented on README.md:
```diff
@@ -98,6 +98,9 @@ To configure options for the builder see the `build_config`
 * `outputStyle`: Supports `expanded` or `compressed`. Defaults to `expanded` in
   dev mode, and `compressed` in release mode.
+* `includePaths`: Supports adding a list of  directories that will be searched 
+  as part of an @include directive. Note that these files must still be readable by
```
nit: Line length

nshahan commented on lib/sass_builder.dart:
```diff
@@ -41,7 +48,7 @@ class SassBuilder implements Builder {
         await buildStep.readAsString(inputId),
         syntax: sass.Syntax.forPath(inputId.path),
-        importers: [new BuildImporter(buildStep)],
+        importers: [new BuildImporter(buildStep, _includePaths)],
```
I'm concerned about this causing errors when your package has a conflicting file with something in the _include paths. For example what if they both contain a file named _mixins.scss at the root. Imports will fail because of the ambiguity.

We could maybe refactor to use a BuildImporter for the package, and then additional BuildImporters for each _includePath. 

Based on the documentation that might cause it to use the first file found which isn't correct but is it more predictable? 
https://github.com/sass/dart-sass/blob/ff3cea5312d5f1f8440aa1fc78dd0a6c3854f993/lib/sass.dart#L38-L49

What do you think? Would you want to experiment with that approach?

fuzzybinary replied:
I would think "first file found", giving preference to relative paths and package: urls would be the correct answer for sure.

I can take a look at creating an IncludePathBuildImporter for sure.

Commit:
Fix documentation nits.

Commit:
Add multiple build importers to search include paths separately.

This allows two modules with the same files to be used without conflict by specifying search order.

fuzzybinary commented:
Should be ready to go!

nshahan commented on lib/src/build_importer.dart:
```diff
@@ -53,12 +54,26 @@ class BuildImporter extends sass.AsyncImporter {
-        p.url.join(p.dirname(import), '_${p.basename(import)}'),
-        from: _buildStep.inputId);
+    AssetId partialId;
+    if (_includePath == null) {
```
It seems that the only difference between the two cases is where the directory is rooted. Could we give a BuildImporter a default root when it is created without an `_includePath` and then keep these as a single codepath?

fuzzybinary replied:
I'll have to think about it, but I think that there are a lot of different possibilities that the first case is dealing with that don't involve a default root. That would include resolution of package: urls and relative imports. The version with _includePath has to attempt to import using "asset" and a non-relative directory. It makes a lot of assumptions I think that the main case doesn't / shouldn't.

With that said I'm not 100% well versed on asset resolution in build_runner so a secondary opinion would be welcome.

nshahan commented on lib/src/build_importer.dart:
```diff
@@ -53,12 +54,26 @@ class BuildImporter extends sass.AsyncImporter {
+    final importId = _includePath == null
+        ? new AssetId.resolve(import, from: _buildStep.inputId)
+        : new AssetId.resolve(p.url.join(
+            'asset:${_buildStep.inputId.package}/${_includePath}', import));
```
Do we need to omit the `from` parameter on these `resolve()` calls?

fuzzybinary replied:
We can omit it because I'm assuming include paths are not relative.  From should only be necessary for relative imports.

fuzzybinary replied:
I ran tests adding the from parameter and they still all pass. I'd have to look for an expect case where they would fail.

nshahan reviewed (commented):
Thanks for taking a look, I left a few comments.

Pardon my confusion here but I'm not very familiar with Sass outside of the Dart ecosystem. I'm still not totally clear how to test out a real world example to see this working. How would I go about using the MDC-web styles that have imports that start with @material?

jifalops commented:
@nshahan I have a couple of example projects you can look at that currently use the one-liner work around for this issue. [Here](https://github.com/jifalops/mdc_web/tree/master/example/demo). Just change that one line in pubspec.yaml to point to your sass_builder and run `npm i --save-dev` from the `web/` directory to install the web components (see package.json). There's also a .scss file that imports the @material stuff and uses some of it.

nshahan commented:
@jifalops Thanks that helps. I'm going to play around with these a little to get a feel for how this can work.
